File: ios/Classes/BannerView/BannerView.swift
//
//  BannerView.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter

private let logTag = "[BannerView]"

class BannerView: NSObject, FlutterPlatformView {
    private let methodHandler: BannerMethodHandler

    private let id: String

    let banner: CASBannerView
    private let bannerCallback: BannerCallback

    init(
        frame: CGRect,
        viewId: Int64,
        args: [String: Any?]?,
        methodHandler: BannerMethodHandler
    ) {
        self.methodHandler = methodHandler

        id = args?["id"] as? String ?? ""
        let contentInfoId = "banner_\(id)"

        let size = if let map = args?["size"] as? [String: Any?] { BannerMethodHandler.getAdSize(map, frame) } else { CASSize.banner }
        banner = CASBannerView(casID: id, size: size)
        banner.tag = Int(viewId)
        let sizeListener = BannerSizeListener(banner, methodHandler, id)
        bannerCallback = BannerCallback(sizeListener, methodHandler, id)
        banner.delegate = bannerCallback

        super.init()

        methodHandler[id] = AdMethodHandler<BannerView>.Ad(ad: self, id: id, contentInfoId: contentInfoId)

        if let args = args {
            if let isAutoloadEnabled = args["isAutoloadEnabled"] as? Bool {
                banner.isAutoloadEnabled = isAutoloadEnabled
            }

            if let refreshInterval = args["refreshInterval"] as? Int {
                banner.refreshInterval = refreshInterval
            }
        } else {
            print("\(logTag) BannerView args is nil")
        }
    }

    func view() -> UIView {
        return banner
    }

    func dispose() {
        _ = methodHandler.remove(id)
        banner.destroy()
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bannerview/BannerView.kt
package com.cleveradssolutions.plugin.flutter.bannerview

import android.content.Context
import android.util.Log
import com.cleveradssolutions.plugin.flutter.bridge.base.AdMethodHandler
import com.cleveradssolutions.plugin.flutter.sdk.OnAdImpressionListenerHandler
import com.cleversolutions.ads.android.CASBannerView
import io.flutter.plugin.platform.PlatformView

private const val LOG_TAG = "CAS.AI.Flutter/BannerView"

class BannerView(
    context: Context,
    viewId: Int,
    args: Map<*, *>?,
    private val methodHandler: BannerMethodHandler
) : PlatformView {

    private val id: String = args?.get("id") as? String ?: ""

    private val banner: CASBannerView

    init {
        val casId = args?.get("casId") as? String
        val contentInfoId = "banner_$id"

        banner = if (casId != null) CASBannerView(context, casId) else CASBannerView(context)
        banner.id = viewId
        val sizeListener = BannerSizeListener(banner, methodHandler, id)
        banner.adListener = BannerCallback(sizeListener, methodHandler, id)
        banner.onImpressionListener =
            OnAdImpressionListenerHandler(methodHandler, id, contentInfoId)

        methodHandler[id] = AdMethodHandler.Ad(this, id, contentInfoId)

        if (args == null) {
            Log.e(LOG_TAG, "BannerView args is null")
        } else {
            (args["size"] as? Map<*, *>)?.let { size ->
                banner.size = BannerMethodHandler.getAdSize(context, size)
            }

            (args["isAutoloadEnabled"] as? Boolean)?.let {
                banner.isAutoloadEnabled = it
            }

            (args["refreshInterval"] as? Int)?.let {
                banner.refreshInterval = it
            }
        }
    }

    override fun getView(): CASBannerView {
        return banner
    }

    override fun dispose() {
        methodHandler.remove(id)
        banner.destroy()
    }

}